classes/CartRule.php

```diff
@@ -668,9 +668,21 @@ public function checkValidity(Context $context, $alreadyInCart = false, $display
 			AND ocr.`id_cart_rule` = ' . (int) $this->id . '
 			AND ' . (int) Configuration::get('PS_OS_ERROR') . ' != o.`current_state`
 			');
-            // When checking the cart rules present in that cart the request result is accurate
-            // When we check if using the cart rule one more time is valid then we increment this value
-            if (!$alreadyInCart) {
+
+            if ($alreadyInCart) {
+                // sometimes a cart rule is already in a cart, but the cart is not yet attached to an order (when logging
+                // in for example), these cart rules are not taken into account buy the query above:
+                // so we count cart rules that are already linked to the current cart but not attached to an order yet.
+                $quantityUsed += (int) Db::getInstance()->getValue('
+                    select count(*)
+                    from ps_cart_cart_rule ccr
+                    LEFT JOIN ps_cart c on c.id_cart = ccr.id_cart
+                    where c.id_customer = ' . $cart->id_customer . ' AND c.id_cart = ' . $cart->id . '
+                    AND NOT EXISTS (SELECT null FROM ps_orders o where o.id_cart = ccr.id_cart)'
```

Comment:
```
                    SELECT null FROM ps_orders o where o.id_cart = ccr.id_cart'
```
It works ? 😄  "SELECT null" ?

Reply:
yes :D 
But I'm working on another fix, I realized this fix "works" but not in all cases

Reply:
I think the fix is good now 👍